spring-integration-jdbc/src/main/java/org/springframework/integration/jdbc/metadata/JdbcMetadataStore.java

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright 2002-2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.jdbc.metadata;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import javax.sql.DataSource;
+
+import org.springframework.dao.EmptyResultDataAccessException;
+import org.springframework.integration.metadata.ConcurrentMetadataStore;
+import org.springframework.integration.metadata.MetadataStore;
+import org.springframework.jdbc.core.JdbcOperations;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.transaction.annotation.Transactional;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+
+/**
+ * Jdbc-based implementation of {@link MetadataStore}.
+ *
+ * @author Bojan Vukasovic
+ * @since 5.0
+ */
+public class JdbcMetadataStore implements ConcurrentMetadataStore {
+
+	/**
+	 * Default value for the table prefix property.
+	 */
+	public static final String DEFAULT_TABLE_PREFIX = "INT_";
+
+	private volatile String tablePrefix = DEFAULT_TABLE_PREFIX;
+
+	private volatile String region = "DEFAULT";
+
+	private final JdbcOperations jdbcTemplate;
+
+	private volatile Map<Query, String> queryCache = new HashMap<>();
+
+	private enum Query {
```

Comment:
One more time (maybe you have just missed my comment):

The approach in the `DefaultLockRepository` looks better for me:
```
public void afterPropertiesSet() {
	this.deleteQuery = String.format(this.deleteQuery, this.prefix);
	this.deleteExpiredQuery = String.format(this.deleteExpiredQuery, this.prefix);
	this.deleteAllQuery = String.format(this.deleteAllQuery, this.prefix);
	this.updateQuery = String.format(this.updateQuery, this.prefix);
	this.insertQuery = String.format(this.insertQuery, this.prefix);
	this.countQuery = String.format(this.countQuery, this.prefix);
}
```
There is no reason to keep lazy-load cache because we have only several entities. And only with variable usage that should be much faster, than map access on the matter.
WDYT?